### PR TITLE
Simplify Mutect2 read filtering

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes issue #672

### CODE REVIEW

**Comment by davidbenjamin** [comment]
@takutosato Not very high priority, but should be quick.  HaplotypeCaller is unaffected, and now Mutect does the same thing as HaplotypeCaller.  Tumor and normal filtering is now the same.  The old idea was to make normal filtering more permissive to retain as much evidence of variation in the normal as possible, but when you think about it we're looking for a variant genotype with allele fraction 1/2 or 1 in the normal, so a few stray alt reads aren't going to matter.  And even if they did, they would *not* be evidence of the variant existing in the normal.  They might be evidence that the *site* is noisy, but that's a separate issue.

**Comment by droazen** [comment]
@davidbenjamin This exact change was PR'd a few months ago, and I requested in the review that the filtering for HC and Mutect be kept separate, so that we can easily adjust them independently, rather than calling into a shared method.

**Comment by davidbenjamin** [comment]
@droazen I would be surprised if we ever needed different read filtering in M2, but we're still experimenting and I see your point.  I will re-separate the methods.  @takutosato there will still be some non-trivial changes here because I'm making the M2 filtering simpler.

**Comment by davidbenjamin** [comment]
Okay, tests are passing again, not sure what Travis was up to.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java [hunk 1]
```diff
@@ -454,26 +452,18 @@ private boolean isNonRef(final byte refBase, final PileupElement p) {
...
-            if (isReadFromNormal(rec)) {
-                if( rec.getLength() < MIN_READ_LENGTH ) {
-                    readsToRemove.add(rec);
-                }
-            } else if( rec.getLength() < MIN_READ_LENGTH || rec.getMappingQuality() < MTAC.MIN_MAPPING_QUALITY_SCORE || hasBadMate(rec) ||
-                    (MTAC.keepRG != null && !rec.getReadGroup().equals(MTAC.keepRG)) ) {
-                readsToRemove.add(rec);
-            }
-        }
+    private Set<GATKRead> filterNonPassingReads( final AssemblyRegion assemblyRegion) {
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java]: also, rename the method to `filterReads`

**Comment by takutosato** [comment]
@davidbenjamin I just pushed my commit. Here are the highlights:

* `hasBadMate()` and `ReadFilterLibrary.MATE_ON_SAME_CONTIG_OR_NO_MAPPED_MATE` are equivalent
* call `regionForGenotyping.removeAll(readsToRemove)` outside of the method to make the side-effect conspicuous
* compose a Predicate from smaller Predicates. I thought it'd be overkill to create a subclass of `ReadFilter` for each criterion.

What do you think?

**Comment by davidbenjamin** [comment]
These are great changes!  I will deal with your previous comments in a new commit on top of yours and submit for re-review.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java [hunk 2]
```diff
@@ -153,6 +150,7 @@ private void initialize() {
         // In Mutect2 we intentionally keep poorly mapped reads for reassembly
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java (hunk 1 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java]: done by you

**Comment by davidbenjamin** [comment]
@takutosato back to you.  I loved the `Predicate` idea and realized that the composed `Predicate` could be a class-level field.  Then the method to identify reads to remove became a one-liner, which I inlined.

**Comment by takutosato** [comment]
awesome! the code looks good. please merge!

**Pull request merged**